platform/linux-generic/odp_crypto_ipsecmb.c

```diff
@@ -0,0 +1,973 @@
+	}
+
+	if (session->p.cipher_key.length == 16) {
+		/* ZUC128 EEA3 */
+		IMB_ZUC_EEA3_1_BUFFER(mb_mgr, session->cipher.key_data,
+				      iv_ptr,
+				      data,
+				      data,
+				      in_len);
+	} else if (session->p.cipher_key.length == 32) {
+		/* ZUC256 EEA3 */
+		IMB_ZUC256_EEA3_1_BUFFER(mb_mgr, session->cipher.key_data,
+					 iv_ptr,
+					 data,
+					 data,
+					 in_len);
+	}
+
+	odp_packet_copy_from_mem(pkt, in_pos, in_len, data);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static int process_zuc_eea3_param(odp_crypto_generic_session_t *session)
+{
+	if (16 != session->p.cipher_key.length &&
+	    32 != session->p.cipher_key.length)
+		return -1;
+
+	if (16 != session->p.cipher_iv_len &&
+	    25 != session->p.cipher_iv_len)
+		return -1;
+
+	memcpy(session->cipher.key_data, session->p.cipher_key.data,
+	       session->p.cipher_key.length);
+
+	session->cipher.func = zuc_eea3_cipher_op;
+
+	return 0;
+}
+
+static
+void auth_zuc_eia3_gen(odp_packet_t pkt,
+		       const odp_crypto_packet_op_param_t *param,
+		       odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->auth_range.offset;
+	uint32_t in_len = param->auth_range.length;
+	uint8_t auth_tag[ZUC_DIGEST_LENGTH];
+
+#if ODP_DEPRECATED_API
+	if (param->auth_iv_ptr) {
+		iv_ptr = param->auth_iv_ptr;
+	} else if (session->p.auth_iv.data) {
+		iv_ptr = session->auth.iv_data;
+	} else {
+		set_crypto_op_result(pkt,
+				     ODP_CRYPTO_ALG_ERR_NONE,
+				     ODP_CRYPTO_ALG_ERR_ICV_CHECK);
+		return;
+	}
+#else
+	iv_ptr = param->auth_iv_ptr;
+	_ODP_ASSERT(session->p.auth_iv_len == 0 || iv_ptr != NULL);
+#endif
+
+	uint32_t seg_len = 0;
+	uint8_t *data = odp_packet_offset(pkt, in_pos, &seg_len, NULL);
+
+	if (odp_unlikely(odp_packet_is_segmented(pkt))) {
+		/* Packet is segmented within the cipher range. Copy the cipher
+		 * range to a contiguous buffer. */
+		odp_packet_copy_to_mem(pkt, in_pos, in_len, local.buffer);
+
+		data = local.buffer;
+	}
+
+	if (session->p.auth_key.length == 16) {
+		/* ZUC128 EIA3 */
+		IMB_ZUC_EIA3_1_BUFFER(mb_mgr, session->auth.key,
+				      iv_ptr,
+				      data,
+				      param->auth_range.length * 8,
+				      (uint32_t *)auth_tag);
+	} else if (session->p.auth_key.length == 32) {
+		/* ZUC256 EIA3 */
+		IMB_ZUC256_EIA3_1_BUFFER(mb_mgr, session->auth.key,
+					 iv_ptr,
+					 data,
+					 param->auth_range.length * 8,
+					 (uint32_t *)auth_tag);
+	}
+
+	/* Copy to the output location */
+	odp_packet_copy_from_mem(pkt, param->hash_result_offset,
+				 session->p.auth_digest_len,
+				 auth_tag);
+
+	set_crypto_op_result_ok(pkt);
+}
+
+static
+void auth_zuc_eia3_check(odp_packet_t pkt,
+			 const odp_crypto_packet_op_param_t *param,
+			 odp_crypto_generic_session_t *session)
+{
+	IMB_MGR *mb_mgr = local.mb_mgr;
+	uint8_t *iv_ptr;
+	uint32_t in_pos = param->auth_range.offset;
+	uint32_t in_len = param->auth_range.length;
+	uint32_t bytes = session->p.auth_digest_len;
```

Comment:
You could use ZUC_DIGEST_LENGTH (or sizeof uint32_t) here to get a bit better compiled result. We know and also check that auth_digest_len is always 4 but the compiler cannot see here that it is effectively a compile time constant.

Reply:
Make sense, updated to ZUC_DIGEST_LENGTH.

Reply:
Now that hash_in and hash:_out are of uint32_t type, you could simlpy compare the variables directly instead of using memcpy.

Reply:
Agree, changed to direct comparison.